Overhaul rank system, command, donator, add probation rank

mheguy opened the pull request:
Overhauls the rank system to allow for expansion/changing quickly and easily.

This enforces a hierarchy of ranks that exists in resources.ranks and everything simply references that table.
features.rank_system hosts the operators to evaluate a player's rank, allows the setting and resetting of ranks, handles the server side of things, and auto-promotes to a new (name-pending) rank between `guest` and `regular`.

In commands, changes the `rank_only` paradigm to instead take the required rank and treats donator as its own distinct flag outside of the hierarchy.

The donator feature is split off from the rankings, and now includes donator_message inside it.

Only mild testing was done in SP, and nothing in MP.

Closes #610
Closes #747

mheguy committed:
Overhaul rank system, command, donator, add probation rank

grilledham commented on map_gen/presets/rotten_apples.lua:
```diff
@@ -194,7 +195,7 @@ local function research_finished(event)
     local force = research.force
 
-    if table.contains(military_techs, research.name) then
+    if array_contains(military_techs, research.name) then
```
Beyond the scope for this PR, but using an array rather than a dictionary for military_techs makes me shudder.

mheguy replied:
Yeah, when I looked at it while making this change I noticed it as well. Time and experience teaches. :D

grilledham commented on utils/table.lua:
```diff
@@ -48,28 +48,47 @@ function table.add_all(t1, t2)
+-- @param t <table>
+-- @param e <any> table element
+-- @returns <number|string|nil> the index of the element or nil
 function table.index_of(t, e)
```
Doesn't this return \<any\>? As you can have any type for the key in a table.

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+--- Check each online player and if their playtime
+local function check_playtime()
+    local auto_trusted = Ranks.auto_trusted
+    local time_for_trust = Config.time_for_trust
```
I don't see this value in the config file.

mheguy replied:
I know I wrote it, so I must have discarded the changes when I discarded debug and cheat mode at some point. :D

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+-- @return <string>
+function Public.get_player_rank_name(player_name)
+    return index_of(Ranks, (player_ranks[player_name] or 0))
+end
```
Given how often this method could be called, I'd be tempted to make a dictionary that maps ranks_number -> rank_name.

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+        Server.set_data(ranking_data_set, player_name, rank)
+        game.print(format("%s set %s's rank to %s.", actor, player_name, Public.get_rank_name(rank)))
+    end
+end
```
Might be worth mentions in the comments to use `reset_rank` for setting to guest

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+
+--- Resets a player's rank
+-- @param player_name <string>
+function Public.reset_rank(player_name)
```
I'd add to the comments that this resets to guest rank and respects auto_trusted time, or words to that effect.

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+
+    for _, p in pairs(game.connected_players) do
+        local player_name = p.name
+        if (p.online_time > time_for_trust) and less_than(player_name, auto_trusted) then
```
Shouldn't this not promote players with the probation rank?

mheguy replied:
Nice catch, thank you.

grilledham commented on features/rank_system.lua:
```diff
@@ -0,0 +1,239 @@
+    end
+)
+
+Event.on_nth_tick(nth_tick, check_playtime)
```
With such a long timeout it would be possible for a player to have:
time_for_trust < player.time < time_for_trust + nth_tick
and not be promoted. For small nth_tick this would be fine, but probably not if it's an hour. 

Another strategy would be to keep track of online guests with player_join and player_leave and only iterate through those players on nth_tick.
Bonus points if you keep it as an ordered list, then you only have to check players until the first fail.

mheguy replied:
Went with a "nothing fancy" approach.

On join and rank == guest enter the `guests` table.

On nth tick
if auto_trusted or higher rank: off the list
elseif guest rank and enough time: off the list and rank to auto_trust
elseif not online: off the list

So you only stay on the list if you are: lower rank than auto_trust, don't have enough time, are online.

If the implementation works for you, feel free to hit resolve convo.

mheguy replied:
This implementation means I can keep the timer at 1h and still catch anyone who meets the criteria since it doesn't rely on a player being online.

mheguy replied:
Because this specifically will be a pain to test in MP, just wanted to show it working in SP:

![image](https://user-images.githubusercontent.com/10162554/52011757-da785100-24a6-11e9-8866-20c7223ed358.png)

grilledham commented on resources/ranks.lua:
```diff
@@ -0,0 +1,7 @@
+    auto_trusted = 1,
+    regular = 2,
+    admin = 3,
+}
```
So these are going to become the values that are stored in Scenario Data?
As changing those values would be annoying, maybe we should use -10, 0, 10, 20, 30 so we have some wiggle to add ranks.

mheguy replied:
Right, to not have to migrate the data as often: that makes total sense.

mheguy replied:
I forgot a couple functions that are important/useful: `increase_player_rank` and decrease. Without being able to add or subtract by 1, I'm actually unsure how else to implement them besides having the ranks be +- 1 of each other.

grilledham replied:
Something like this
```lua
local ranks = {
    probation = -10,
    guest = 0,
    auto_trusted = 10,
    regular = 20,
    admin = 30
}

local sorted_ranks = {}
local rank_to_index = {}

for k, v in pairs(ranks) do
    sorted_ranks[#sorted_ranks + 1] = v
end

table.sort(sorted_ranks)

for k, v in pairs(sorted_ranks) do
    rank_to_index[v] = k
end

function change_rank(current_rank, change)
    local index = rank_to_index[current_rank]

    local new_index = math.clamp(index + change, 1, #sorted_ranks)

    return sorted_ranks[new_index]
end
```
Example of use
```lua
change_rank(ranks.guest,3) -- returns 30
change_rank(ranks.regular,-2) -- returns 0
```

mheguy replied:
Ty for doing my work for me. <3

grilledham commented:
I like this. We will need to test it in MP before I can approve it though.

grilledham commented:
I don't think `set_rank` and `reset_rank` should print messages, that's up to the calling code to decide.
Those messages make sense if the functions are called from a command. But maybe for example the antigreif module wants to call those function, then it would want to suppress the messages, or print different ones.

grilledham commented:
Currently we can set and reset the ranks of players who have never been in the game, I'm ok with that (I'm also ok with not allowing it), I just want to know if that was intentional

plague006 commented:
> Currently we can set and reset the ranks of players who have never been in the game, I'm ok with that (I'm also ok with not allowing it), I just want to know if that was intentional

Intentional as there have been multiple times I wanted to regular or un-regular players who hadn't been on a particular map. But on reflection I should include a warning in those cases since typos are a thing.

plague006 commented:
> 
> 
> > Currently we can set and reset the ranks of players who have never been in the game, I'm ok with that (I'm also ok with not allowing it), I just want to know if that was intentional
> 
> Intentional as there have been multiple times I wanted to regular or un-regular players who hadn't been on a particular map. But on reflection I should include a warning in those cases since typos are a thing.

So that was my view on the matter when I was just a moderator.. as a dev looking at my system and trying to account for typos: I want to enforce the existence of a player when running `/commands` but not have the rank_system module itself check for a LuaPlayer.

mheguy committed:
improve table doc
add instructions for changing ranks in ranks
add missing config entry
refine regular and probation commands

mheguy committed:
add table to lookup name-rank relations
add local function to get player rank
add increase and decrease player rank functions
ad debug-only data migration function

mheguy committed:
Make autopromotion more robust

mheguy committed:
Fix online -> connected

mheguy committed:
Implement grilled maths

mheguy committed:
Many fixes and tweaks

mheguy committed:
Change checks from 1 hr to 15 mins

mheguy committed:
Drop table entries for invalid or non-existent players.

mheguy committed:
Change to index -> luaplayer dictionary for guests

mheguy committed:
Add timezone to redmew_version

mheguy committed:
Add get_donators function

mheguy committed:
Rename functions to be homogeneous

mheguy committed:
Add admin check to get_player_rank

mheguy committed:
Fix data type in doc

mheguy committed:
Add space between time and UTC

mheguy committed:
Add timestamp to log

mheguy committed:
Minor cleanup of player_color

mheguy committed:
tag_group: deafult - > default, remove ipairs

plague006 commented:
redmew_qol.pick_name() will need a fix when `global.regulars` is no longer available

plague006 commented:
Fix conflicts -> rebase and FP

mheguy committed:
Change admin_only to use rank

mheguy committed:
Fix outpost_builder, fix redmew_qol.
Add functions to donator and rank system to get full player tables.

mheguy committed:
Fix player_create ref to usergroups and consolidate conditions. Autoformat.

mheguy committed:
Clean up /whois command

muppet9010 committed:
Add everyone is regular option

plague006 commented:
#747 Added in

mheguy committed:
Improve doc

mheguy committed:
On server start, clear table before writing it when syncing

plague006 commented:
https://github.com/Refactorio/RedMew/pull/713/commits/d21ad8d8b542da0d18eac2c074ec966499d858fe should keep things more consistent so that loaded saves and server scenario data do not conflict.

plague006 commented:
More changes! (yay~~)

Added `increase_player_rank_to` and `decrease_player_rank_to`. Unlike `set_rank` this ensures that when we mean to increase/decrease a rank, we don't accidentally do the opposite.

Remove all prints from the module and instead return strings that calls can choose to use or ignore.

Have `reset_player_rank` call `set_player_rank` to consolidate where we set server data.
---
Add color to `Game.player_print`
---
Overhaul `regular` and `probation` to take format `/regular <add|remove> <name>`.
As part of the overhaul, have them use `increase/decrease_player_rank_to` to restrain their actions.

---

Limited testing done in SP.

mheguy committed:
Add more functions for increasing/decreasing ranks, remove all prints

plague006 commented:
`/whois` 1459.479 Script @/factorio/4/temp/currently-playing/utils/command.lua:253: Error while running 'whois' with arguments {inventory = false, player = "plague006"}: /factorio/4/temp/currently-playing/features/rank_system.lua:80: bad argument #2 to '__index' (string expected, got table)

mheguy committed:
Fix whois sending LuaPlayer instead of name

mheguy committed:
That is *really* the wrong place to print that message

mheguy committed:
Fix no color to player_print

mheguy committed:
Prevent making a player below guest rank a regular

mheguy committed:
Switch info to use locale for donator rank. Make donator a supplement to rank rather than rank itself.

mheguy committed:
Expose get_rank, localize functions that are called frequently

mheguy committed:
Use locale for new rank system

plague006 commented:
RB+FP for merge conflict.
Moved what I could in rank_system into locale, and only what I had to in dependent modules.
Listing of `rank` in `/whois` is disabled until I can figure out a sane way to incorporate all of that together.

plague006 commented:
Best way to tackle whois might be concatenating everything into print, example:
`game.print({"", {"item-name.iron-plate"}, ": ", 60})`

grilledham committed:
change how rank colors are returned

grilledham committed:
updates to player list to show rank and donator

grilledham commented:
I updated the player list gui
![image](https://user-images.githubusercontent.com/10003572/52712382-55a03380-2f8c-11e9-8fff-23d5810202bc.png)
![image](https://user-images.githubusercontent.com/10003572/52712394-5e910500-2f8c-11e9-8028-760da915d3f8.png)
![image](https://user-images.githubusercontent.com/10003572/52712418-6781d680-2f8c-11e9-8f04-3c98c7f33e54.png)
Not tested in multiplayer

mheguy committed:
Restore functionality of whois

mheguy committed:
Check for nil in syn_ranks_callback

mheguy committed:
Fix lint warnings

grilledham reviewed (approved):
On the basis that this was tested in multiplayer, this is probably safe enough to merge.
